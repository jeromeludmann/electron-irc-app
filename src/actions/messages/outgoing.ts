import { RoutedAction, BufferKey } from '@app/utils/Route'

export const SEND_MESSAGE = 'MESSAGE/SEND_MESSAGE'

export interface SendMessageAction<T = string, P = {}> extends RoutedAction<T> {
  group: typeof SEND_MESSAGE
  payload: { raw: string } & P
}

export const SEND_JOIN = 'MESSAGE/SEND_JOIN'

export type SendJoinAction = SendMessageAction<
  typeof SEND_JOIN,
  { channel: string }
>

export function sendJoin(serverKey: string, channel: string): SendJoinAction {
  return {
    type: SEND_JOIN,
    group: SEND_MESSAGE,
    payload: { channel, raw: raw('join', channel) },
    route: { serverKey, bufferKey: BufferKey.NONE },
  }
}

export const SEND_NICK = 'MESSAGE/SEND_NICK'

export type SendNickAction = SendMessageAction<
  typeof SEND_NICK,
  { nick: string }
>

export function sendNick(serverKey: string, nick: string): SendNickAction {
  return {
    type: SEND_NICK,
    group: SEND_MESSAGE,
    payload: { nick, raw: raw('nick', nick) },
    route: { serverKey, bufferKey: BufferKey.NONE },
  }
}

export const SEND_PART = 'MESSAGE/SEND_PART'

export type SendPartAction = SendMessageAction<
  typeof SEND_PART,
  { channel: string; text: string }
>

export function sendPart(
  serverKey: string,
  channel: string,
  text: string = '',
): SendPartAction {
  return {
    type: SEND_PART,
    group: SEND_MESSAGE,
    payload: { channel, text, raw: raw('part', channel, text) },
    route: { serverKey, bufferKey: BufferKey.NONE },
  }
}

export const SEND_PING_TO_SERVER = 'MESSAGE/SEND_PING_TO_SERVER'

export type SendPingToServerAction = SendMessageAction<
  typeof SEND_PING_TO_SERVER,
  { key: number }
>

export function sendPingToServer(
  serverKey: string,
  key = Date.now(),
): SendPingToServerAction {
  return {
    type: SEND_PING_TO_SERVER,
    group: SEND_MESSAGE,
    payload: { key, raw: raw('ping', key.toString()) },
    route: { serverKey, bufferKey: BufferKey.STATUS },
  }
}

export const SEND_PONG_TO_SERVER = 'MESSAGE/SEND_PONG_TO_SERVER'

export type SendPongToServerAction = SendMessageAction<
  typeof SEND_PONG_TO_SERVER,
  { key: string }
>

export function sendPongToServer(
  serverKey: string,
  key: string,
): SendPongToServerAction {
  return {
    type: SEND_PONG_TO_SERVER,
    group: SEND_MESSAGE,
    payload: { key, raw: raw('pong', key) },
    route: { serverKey, bufferKey: BufferKey.STATUS },
  }
}

export const SEND_PRIVMSG = 'MESSAGE/SEND_PRIVMSG'

export type SendPrivmsgAction = SendMessageAction<
  typeof SEND_PRIVMSG,
  { channel: string; text: string }
>

export function sendPrivmsg(
  serverKey: string,
  channel: string,
  text: string,
): SendPrivmsgAction {
  return {
    type: SEND_PRIVMSG,
    group: SEND_MESSAGE,
    payload: { channel, text, raw: raw('privmsg', channel, text) },
    route: { serverKey, bufferKey: channel },
  }
}

export const SEND_QUIT = 'MESSAGE/SEND_QUIT'

export type SendQuitAction = SendMessageAction<
  typeof SEND_QUIT,
  { text: string }
>

export function sendQuit(serverKey: string, text: string = ''): SendQuitAction {
  return {
    type: SEND_QUIT,
    group: SEND_MESSAGE,
    payload: { text, raw: raw('quit', text) },
    route: { serverKey, bufferKey: BufferKey.NONE },
  }
}

export const SEND_USER = 'MESSAGE/SEND_USER'

export type SendUserAction = SendMessageAction<
  typeof SEND_USER,
  { username: string; realname: string }
>

export function sendUser(
  serverKey: string,
  username: string,
  realname: string,
): SendUserAction {
  return {
    type: SEND_USER,
    group: SEND_MESSAGE,
    payload: {
      username,
      realname,
      raw: raw('user', username, '0', '*', realname),
    },
    route: { serverKey, bufferKey: BufferKey.NONE },
  }
}

function raw(command: string, ...params: string[]) {
  const last = params.length - 1

  if (params[last].indexOf(' ') > -1 || params[last].charAt(0) === ':') {
    params[last] = `:${params[last]}`
  }

  return `${command.toUpperCase()} ${params.join(' ')}`.trimRight()
}
