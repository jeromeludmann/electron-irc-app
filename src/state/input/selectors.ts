import { createSelector } from 'reselect'
import { getBufferFactory } from '@app/state/buffer/selectors'
import { Route } from '@app/utils/Route'

export function getInput(route?: Route) {
  return createSelector(getBufferFactory(route), channel => channel.input)
}

export function getInputValue(route?: Route) {
  return createSelector(getInput(route), input => input.value)
}
