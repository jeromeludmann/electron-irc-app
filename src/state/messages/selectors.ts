import { createSelector } from 'reselect'
import { getBufferFactory } from '@app/state/buffer/selectors'
import { Route } from '@app/utils/Route'

export function getMessages(route?: Route) {
  return createSelector(getBufferFactory(route), buffer => buffer.messages)
}
