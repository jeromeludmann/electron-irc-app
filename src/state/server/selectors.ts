import { createSelector } from 'reselect'
import { getServers } from '@app/state/root/selectors'
import { getRoute } from '@app/state/route/selectors'

export function getServerFactory(serverKey?: string) {
  return createSelector(
    getServers,
    getRoute,
    (servers, route) => servers[serverKey || route.serverKey],
  )
}

export function getBuffers(serverKey?: string) {
  return createSelector(getServerFactory(serverKey), server => server.buffers)
}
