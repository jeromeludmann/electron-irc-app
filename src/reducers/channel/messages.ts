import { Action } from "redux";
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
} from "@app/actions/socket";
import {
  ReceiveErrorAction,
  ReceiveJoinAction,
  ReceivePartAction,
  ReceivePrivmsgAction,
  ReceiveNoticeFromServerAction,
  ReceiveNoticeFromChannelAction,
  ReceiveNoticeFromUserAction,
  RECEIVE_ERROR,
  RECEIVE_JOIN,
  RECEIVE_NOTICE_FROM_SERVER,
  RECEIVE_NOTICE_FROM_CHANNEL,
  RECEIVE_NOTICE_FROM_USER,
  RECEIVE_PART,
  RECEIVE_PRIVMSG,
  RECEIVE_PING_FROM_SERVER,
  RECEIVE_PONG_FROM_SERVER,
  ReceivePongFromServerAction,
} from "@app/actions/msgIncoming";
import {
  SEND_PONG_TO_SERVER,
  SEND_PRIVMSG,
  SendPrivmsgAction,
} from "@app/actions/msgOutgoing";
import {
  PRINT_HELP_BY_DEFAULT,
  PrintHelpByDefaultAction,
  PRINT_HELP_ABOUT_COMMAND,
  PrintHelpAboutCommandAction,
} from "@app/actions/commands";
import { RouteState } from "@app/reducers/route";
import { UserState } from "@app/reducers/server/user";
import { RoutedAction } from "@app/Route";

// TODO replace string[] by MessageState[]
export type MessagesState = Readonly<string[]>;

export const messagesInitialState: MessagesState = [];

const handlers: {
  [action: string]: (
    messages: MessagesState,
    action: Action,
    extraStates: { route: RouteState; user: UserState },
  ) => MessagesState;
} = {
  [CONNECTION_CLOSED]: messages => [
    ...messages,
    "Disconnected from remote host.",
  ],

  [CONNECTION_FAILED]: (
    messages,
    { payload: { message } }: ConnectionFailedAction,
  ) => [...messages, message],

  [RECEIVE_ERROR]: (messages, { payload: { message } }: ReceiveErrorAction) => [
    ...messages,
    message,
  ],

  [RECEIVE_JOIN]: (
    messages,
    { payload: { user, channel } }: ReceiveJoinAction,
  ) => [...messages, `${user.nick} has joined ${channel}`],

  [RECEIVE_NOTICE_FROM_SERVER]: (
    messages,
    action: ReceiveNoticeFromServerAction,
  ) => [...messages, action.payload.text],

  [RECEIVE_NOTICE_FROM_CHANNEL]: (
    messages,
    {
      payload: { user, text },
      route: { channelKey },
    }: ReceiveNoticeFromChannelAction,
  ) => [...messages, `-${user.nick}/${channelKey}- ${text}`],

  [RECEIVE_NOTICE_FROM_USER]: (
    messages,
    { payload: { user, text } }: ReceiveNoticeFromUserAction,
  ) => [...messages, `-${user.nick}- ${text}`],

  [RECEIVE_PART]: (
    messages,
    { payload: { user, channel, message } }: ReceivePartAction,
  ) => {
    const baseMsg = `${user.nick} has left ${channel}`;
    return [...messages, message ? `${baseMsg} (${message})` : baseMsg];
  },

  [RECEIVE_PONG_FROM_SERVER]: (
    messages,
    { payload: { key, lag } }: ReceivePongFromServerAction,
    {},
  ) => [...messages, `Response from server ${key} in ${lag} ms`],

  [PRINT_HELP_BY_DEFAULT]: (
    messages,
    { payload: { commands } }: PrintHelpByDefaultAction,
  ) => [
    ...messages,
    ...Object.keys(commands).map(
      commandName => `${commandName} - ${commands[commandName].description}`,
    ),
    "Type /HELP <command> for more details.",
  ],

  [PRINT_HELP_ABOUT_COMMAND]: (
    messages,
    { payload: { command } }: PrintHelpAboutCommandAction,
  ) => [
    ...messages,
    `Usage: /${command.name} ${command.syntax} - ${command.description}`,
  ],

  [RECEIVE_PRIVMSG]: (
    messages,
    { payload: { user, text } }: ReceivePrivmsgAction,
  ) => [...messages, `${user.nick}: ${text}`],

  [RECEIVE_PING_FROM_SERVER]: messages => [...messages, "Ping?"],

  [RAW_MESSAGES_RECEIVED]: (messages, action: RawMessagesAction) => [
    ...messages,
    ...action.payload.messages,
  ],

  [SEND_PONG_TO_SERVER]: messages => [...messages, "Pong!"],

  [SEND_PRIVMSG]: (messages, action: SendPrivmsgAction, { user: { nick } }) => [
    ...messages,
    `${nick}: ${action.payload.text}`,
  ],
};

export const reduceMessages = (
  messagesState = messagesInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState; user: UserState },
) =>
  handlers.hasOwnProperty(action.type)
    ? handlers[action.type](messagesState, action, extraStates)
    : messagesState;
