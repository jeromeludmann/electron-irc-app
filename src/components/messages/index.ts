import { connect, MapStateToProps } from "react-redux";
import MessageList from "@app/components/messages/MessageList";
import { MessagesState } from "@app/reducers/channel/messages";
import { AppState } from "@app/reducers";
import { selectMessages } from "@app/reducers/channel/_selectors";

interface StateProps {
  messages: MessagesState;
}

const mapStateToProps: MapStateToProps<StateProps, {}, AppState> = state => ({
  messages: selectMessages(state),
});

export default connect(mapStateToProps)(MessageList);
