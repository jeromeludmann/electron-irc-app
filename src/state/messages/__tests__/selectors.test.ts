import { getMessages } from '@app/state/messages/selectors'
import { rootInitialState } from '@app/state/root/reducer'

describe('getMessages()', () => {
  it('should get messages', () => {
    expect(getMessages()(rootInitialState)).toMatchSnapshot()
  })
})
