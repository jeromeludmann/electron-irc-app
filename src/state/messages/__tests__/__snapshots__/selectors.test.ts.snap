// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMessages() should get messages 1`] = `Array []`;
